lightning/src/ln/features.rs

```diff
@@ -8,45 +23,253 @@ use std::marker::PhantomData;
-	impl Context for ChannelContext {}
-
-	pub trait DataLossProtect: Context {}
-	impl DataLossProtect for InitContext {}
-	impl DataLossProtect for NodeContext {}
+mod sealed {
+	/// The context in which [`Features`] are applicable. Defines which features are required and
+	/// which are optional for the context.
+	///
+	/// [`Features`]: ../struct.Features.html
+	pub trait Context {
+		/// Features that are known to the implementation, where a required feature is indicated by
+		/// its even bit and an optional feature is indicated by its odd bit.
+		const KNOWN_FEATURE_FLAGS: &'static [u8];
+
+		/// Bitmask for selecting features that are known to the implementation, regardless of
+		/// whether each feature is required or optional.
+		const KNOWN_FEATURE_MASK: &'static [u8];
+	}
+
+	/// Defines a [`Context`] by stating which features it requires and which are optional. Features
+	/// are specified as a comma-separated list of bytes where each byte is a pipe-delimited list of
+	/// feature identifiers.
+	///
+	/// [`Context`]: trait.Context.html
+	macro_rules! define_context {
+		($context: ident {
+			required_features: [$( $( $required_feature: ident )|*, )*],
+			optional_features: [$( $( $optional_feature: ident )|*, )*],
+		}) => {
+			pub struct $context {}
+
+			impl Context for $context {
+				const KNOWN_FEATURE_FLAGS: &'static [u8] = &[
+					// For each byte, use bitwise-OR to compute the applicable flags for known
+					// required features `r_i` and optional features `o_j` for all `i` and `j` such
+					// that the following slice is formed:
+					//
+					// [
+					//  `r_0` | `r_1` | ... | `o_0` | `o_1` | ...,
+					//  ...,
+					// ]
+					$(
+						0b00_00_00_00 $(|
+							<Self as $required_feature>::REQUIRED_MASK)*
+						$(|
+							<Self as $optional_feature>::OPTIONAL_MASK)*,
+					)*
+				];
+
+				const KNOWN_FEATURE_MASK: &'static [u8] = &[
+					// Similar as above, but set both flags for each feature regardless of whether
+					// the feature is required or optional.
+					$(
+						0b00_00_00_00 $(|
+							<Self as $required_feature>::REQUIRED_MASK |
+							<Self as $required_feature>::OPTIONAL_MASK)*
+						$(|
+							<Self as $optional_feature>::REQUIRED_MASK |
+							<Self as $optional_feature>::OPTIONAL_MASK)*,
+					)*
+				];
+			}
+		};
+	}
+
+	define_context!(InitContext {
+		required_features: [
+			// Byte 0
+			,
+			// Byte 1
+			,
+			// Byte 2
+			,
+		],
+		optional_features: [
+			// Byte 0
+			DataLossProtect | InitialRoutingSync | UpfrontShutdownScript,
+			// Byte 1
+			VariableLengthOnion | PaymentSecret,
+			// Byte 2
+			BasicMPP,
+		],
+	});
+	define_context!(NodeContext {
+		required_features: [
+			// Byte 0
+			,
+			// Byte 1
+			,
+			// Byte 2
+			,
+		],
+		optional_features: [
+			// Byte 0
+			DataLossProtect | UpfrontShutdownScript,
+			// Byte 1
+			VariableLengthOnion | PaymentSecret,
+			// Byte 2
+			BasicMPP,
+		],
+	});
+	define_context!(ChannelContext {
```

Comment:
Yes I think this works correctly even with `C+` distinction for `option_support_large_channel`, I mean it's up to us to encode as required features only at context declaration.